### DISCUSSION TITLE
NODE_CONFIG_DIR is not updating when using import

### DISCUSSION THREAD

**Discussion by vinayverismart** [node-config/node-config#750] [Q&A] [answered]
I am trying to update NODE_CONFIG_DIR 
But when I am using `require`
```
process.env["NODE_CONFIG_DIR"] = process.cwd() + "/test/config/";
delete require.cache[require.resolve('config')];
var config = require('config');
console.log('NODE_CONFIG_DIR: ' + config.util.getEnv('NODE_CONFIG_DIR')); 
```

Output is `../test/config` (which is correct)
But when I am using `import`
```
process.env["NODE_CONFIG_DIR"] = process.cwd() + "/test/config/";
import config from "config";
console.log('NODE_CONFIG_DIR: ' + config.util.getEnv('NODE_CONFIG_DIR')); 
```
Output is `../config` (It should be `../test/config`)
And if I am trying to use `config.get()`, it is returning the value from config which is located in `../config`

When I am printing `process.env["NODE_CONFIG_DIR"]`, path is `../test/config` in both cases

**Comment by yiqingpeng** [comment]
I encounter the same issue. Why? The bug is obvious, but the author does not care it?

**Reply by yiqingpeng** [reply]
import statement is called first, so process.env["NODE_CONFIG_DIR"] = process.cwd() + "/test/config/"; will be executed after all import statements. The author should consider this case.

**Comment by yiqingpeng** [comment]
@vinayverismart You can use import function like this: `const config = (await import("config")).default;` to replace import statement.

**Comment by jdmarshall** [comment]
And this is the other reason why my attempts to convert this package to ESM stalled. 

Import cannot be forced to reload a package the way require can. And all of the test suites rely on doing so. 

There’s also another problem with import and using it inside the config directory that I got wedged on, which is the more intractable issue.

**Reply by jdmarshall** [reply]
There is a way:  

https://nodejs.org/api/esm.html#file-urls

But this seems more useful for test suites than production code.

**Comment by jdmarshall** [comment, marked as answer]
You will need to set these environment variables before triggering the import chain because `import` is far more aggressive than `require()`. This is a Node.js decision about determinism rather than a bug with node-config.

Edit to add: There aren't a lot of environment variables I like to set in a setup shell or in pm2.config but this is for certain one of them.
